homeassistant/components/prometheus/sensor.py

```diff
@@ -0,0 +1,114 @@
+"""Prometheus Sensor component."""
+from datetime import timedelta
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA as SENSOR_PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_NAME,
+    CONF_UNIT_OF_MEASUREMENT,
+    CONF_URL,
+    STATE_UNKNOWN,
+)
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+
+DEFAULT_URL = "http://localhost:9090"
+CONF_QUERIES = "queries"
+CONF_EXPR = "expr"
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=60)
+
+_LOGGER = logging.getLogger(__name__)
+
+_QUERY_SCHEMA = vol.Schema(
+    {
+        vol.Required(
+            CONF_NAME,
+        ): cv.string,
+        vol.Optional(CONF_UNIT_OF_MEASUREMENT): cv.string,
+        vol.Required(CONF_EXPR): cv.string,
+    }
+)
+
+PLATFORM_SCHEMA = SENSOR_PLATFORM_SCHEMA.extend(
+    {
+        vol.Optional(CONF_URL, default=DEFAULT_URL): cv.string,
+        vol.Required(CONF_QUERIES): [_QUERY_SCHEMA],
+    }
+)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the sensor platform."""
+    session = async_get_clientsession(hass)
+
+    prometheus = Prometheus(config.get(CONF_URL), session)
+
+    sensors = []
+    for query in config.get("queries", dict()):
+        sensors.append(PrometheusSensor(prometheus, query))
+
+    async_add_entities(sensors, update_before_add=True)
+
+
+class Prometheus:
+    """Wrapper for Prometheus API Requests."""
+
+    def __init__(self, url, session):
+        """Initialize the Prometheus API wrapper."""
+        self._session = session
+        self._url = f"{url}/api/v1/query"
```

Comment:
I'm not really happy about using string formatting to build a URL, but `urllib.parse.urljoin` won't work for `https://monitoring.example.com/prometheus`, as it would override `/override` with `/api/v1/query` instead of appending to it.

Reply:
Won't `urljoin(f"{url}/", "api/v1/query")` work for you here?

Reply:
```
>>> urljoin("http://monitoring.example.com/prometheus", "/api/v1/query")
'http://monitoring.example.com/api/v1/query'
```

I strips the `/prometheus` base path.

Reply:
Note that I am suggesting to append a slash to `url` and add `api/v1/query` as a relative path.

Reply:
```
>>> urljoin("http://monitoring.example.com/prometheus/", "/api/v1/query")
'http://monitoring.example.com/api/v1/query'
```

Like this?

Reply:
```
>>> url = "http://monitoring.example.com/prometheus"
>>> urljoin(f"{url}/", "api/v1/query")
'http://monitoring.example.com/prometheus/api/v1/query'
```

Reply:
Ah, so the problem was the path being relative. Thanks!

Reply:
Applied.